sql: TestDescriptorRefreshOnRetry failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/20d8bd495a0919f70479b031161b18632a2445f4

Parameters:
```
TAGS=
GOFLAGS=
```

Stress build found a failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=627894&tab=buildLog

```
W180427 10:24:36.484340 62873 server/status/runtime.go:109  Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I180427 10:24:36.489524 62873 server/server.go:763  [n?] monitoring forward clock jumps based on server.clock.forward_jump_check_enabled
I180427 10:24:36.511946 62873 server/config.go:546  [n?] 1 storage engine initialized
I180427 10:24:36.511964 62873 server/config.go:549  [n?] RocksDB cache size: 128 MiB
I180427 10:24:36.511971 62873 server/config.go:549  [n?] store 0: in-memory, size 0 B
I180427 10:24:36.515821 62873 server/node.go:374  [n?] **** cluster 3db1b2e8-6dce-4092-ad5f-cf139ee729c1 has been created
I180427 10:24:36.515874 62873 server/server.go:1333  [n?] **** add additional nodes by specifying --join=127.0.0.1:39113
I180427 10:24:36.517088 62873 storage/store.go:1385  [n1,s1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I180427 10:24:36.517332 62873 server/node.go:512  [n1] initialized store [n1,s1]: disk (capacity=512 MiB, available=512 MiB, used=0 B, logicalBytes=6.9 KiB), ranges=1, leases=0, writes=0.00, bytesPerReplica={p10=7043.00 p25=7043.00 p50=7043.00 p75=7043.00 p90=7043.00 pMax=7043.00}, writesPerReplica={p10=0.00 p25=0.00 p50=0.00 p75=0.00 p90=0.00 pMax=0.00}
I180427 10:24:36.517380 62873 server/node.go:352  [n1] node ID 1 initialized
I180427 10:24:36.517445 62873 gossip/gossip.go:332  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:39113" > attrs:<> locality:<> ServerVersion:<major_val:2 minor_val:0 patch:0 unstable:3 > 
I180427 10:24:36.520440 62873 storage/stores.go:309  [n1] read 0 node addresses from persistent storage
I180427 10:24:36.520514 62873 server/node.go:653  [n1] connecting to gossip network to verify cluster ID...
I180427 10:24:36.520534 62873 server/node.go:678  [n1] node connected via gossip and verified as part of cluster "3db1b2e8-6dce-4092-ad5f-cf139ee729c1"
I180427 10:24:36.520553 62873 server/node.go:446  [n1] node=1: started with [<no-attributes>=<in-mem>] engine(s) and attributes []
I180427 10:24:36.520633 62873 server/server.go:1459  [n1] starting https server at 127.0.0.1:46239
I180427 10:24:36.520648 62873 server/server.go:1460  [n1] starting grpc/postgres server at 127.0.0.1:39113
I180427 10:24:36.520661 62873 server/server.go:1461  [n1] advertising CockroachDB node at 127.0.0.1:39113
W180427 10:24:36.520674 62873 sql/jobs/registry.go:286  [n1] unable to get node liveness: node not in the liveness table
I180427 10:24:36.529640 62956 storage/replica_command.go:862  [split,n1,s1,r1/1:/M{in-ax}] initiating a split of this range at key /System/"" [r2]
I180427 10:24:36.532466 63156 storage/replica_command.go:862  [split,n1,s1,r2/1:/{System/-Max}] initiating a split of this range at key /System/NodeLiveness [r3]
I180427 10:24:36.540075 62917 storage/replica_command.go:862  [split,n1,s1,r3/1:/{System/NodeL…-Max}] initiating a split of this range at key /System/NodeLivenessMax [r4]
I180427 10:24:36.543947 62919 storage/replica_command.go:862  [split,n1,s1,r4/1:/{System/NodeL…-Max}] initiating a split of this range at key /System/tsd [r5]
I180427 10:24:36.547769 62922 storage/replica_command.go:862  [split,n1,s1,r5/1:/{System/tsd-Max}] initiating a split of this range at key /System/"tse" [r6]
I180427 10:24:36.552243 62925 storage/replica_command.go:862  [split,n1,s1,r6/1:/{System/tse-Max}] initiating a split of this range at key /Table/SystemConfigSpan/Start [r7]
I180427 10:24:36.555719 63189 storage/replica_command.go:862  [split,n1,s1,r7/1:/{Table/System…-Max}] initiating a split of this range at key /Table/11 [r8]
I180427 10:24:36.559887 63197 storage/replica_command.go:862  [split,n1,s1,r8/1:/{Table/11-Max}] initiating a split of this range at key /Table/12 [r9]
I180427 10:24:36.563773 63174 storage/replica_command.go:862  [split,n1,s1,r9/1:/{Table/12-Max}] initiating a split of this range at key /Table/13 [r10]
I180427 10:24:36.566818 63182 storage/replica_command.go:862  [split,n1,s1,r10/1:/{Table/13-Max}] initiating a split of this range at key /Table/14 [r11]
I180427 10:24:36.570739 63165 storage/replica_command.go:862  [split,n1,s1,r11/1:/{Table/14-Max}] initiating a split of this range at key /Table/15 [r12]
I180427 10:24:36.576666 63209 storage/replica_command.go:862  [split,n1,s1,r12/1:/{Table/15-Max}] initiating a split of this range at key /Table/16 [r13]
I180427 10:24:36.581964 62873 sql/event_log.go:121  [n1] Event: "set_cluster_setting", target: 0, info: {SettingName:diagnostics.reporting.enabled Value:true User:node}
I180427 10:24:36.583597 63266 storage/replica_command.go:862  [split,n1,s1,r13/1:/{Table/16-Max}] initiating a split of this range at key /Table/17 [r14]
W180427 10:24:36.594979 63240 storage/intent_resolver.go:320  [n1,s1] failed to push during intent resolution: failed to push "split" id=1fd86af1 key=/Local/Range/Table/16/RangeDescriptor rw=true pri=0.01492840 iso=SERIALIZABLE stat=PENDING epo=0 ts=1524824676.583607762,0 orig=1524824676.583607762,0 max=1524824676.583607762,0 wto=false rop=false seq=3
I180427 10:24:36.597246 63120 storage/replica_command.go:862  [split,n1,s1,r14/1:/{Table/17-Max}] initiating a split of this range at key /Table/18 [r15]
I180427 10:24:36.604370 62873 sql/event_log.go:121  [n1] Event: "set_cluster_setting", target: 0, info: {SettingName:version Value:$1 User:node}
I180427 10:24:36.608260 63288 storage/replica_command.go:862  [split,n1,s1,r15/1:/{Table/18-Max}] initiating a split of this range at key /Table/19 [r16]
I180427 10:24:36.608849 62873 sql/event_log.go:121  [n1] Event: "set_cluster_setting", target: 0, info: {SettingName:trace.debug.enable Value:false User:node}
I180427 10:24:36.616664 63316 storage/replica_command.go:862  [split,n1,s1,r16/1:/{Table/19-Max}] initiating a split of this range at key /Table/20 [r17]
I180427 10:24:36.620159 62873 sql/event_log.go:121  [n1] Event: "set_cluster_setting", target: 0, info: {SettingName:cluster.secret Value:gen_random_uuid()::STRING User:node}
I180427 10:24:36.620954 62873 server/server.go:1539  [n1] done ensuring all necessary migrations have run
I180427 10:24:36.620970 62873 server/server.go:1542  [n1] serving sql connections
I180427 10:24:36.621014 63357 storage/replica_command.go:862  [split,n1,s1,r17/1:/{Table/20-Max}] initiating a split of this range at key /Table/21 [r18]
I180427 10:24:36.623645 63326 sql/event_log.go:121  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:127.0.0.1:39113} Attrs: Locality: ServerVersion:2.0-3} ClusterID:3db1b2e8-6dce-4092-ad5f-cf139ee729c1 StartedAt:1524824676520537462 LastUp:1524824676520537462}
I180427 10:24:36.625622 63365 storage/replica_command.go:862  [split,n1,s1,r18/1:/{Table/21-Max}] initiating a split of this range at key /Table/22 [r19]
I180427 10:24:36.630224 63378 storage/replica_command.go:862  [split,n1,s1,r19/1:/{Table/22-Max}] initiating a split of this range at key /Table/23 [r20]
I180427 10:24:36.647594 63279 sql/event_log.go:121  [n1,client=127.0.0.1:33874,user=root] Event: "create_database", target: 50, info: {DatabaseName:t Statement:CREATE DATABASE t User:root}
I180427 10:24:36.647986 63280 storage/replica_command.go:862  [split,n1,s1,r20/1:/{Table/23-Max}] initiating a split of this range at key /Table/50 [r21]
I180427 10:24:36.650459 63279 sql/event_log.go:121  [n1,client=127.0.0.1:33874,user=root] Event: "create_table", target: 51, info: {TableName:t.public.foo Statement:CREATE TABLE t.public.foo (v INT) User:root}
I180427 10:24:36.651186 63337 storage/replica_command.go:862  [split,n1,s1,r21/1:/{Table/50-Max}] initiating a split of this range at key /Table/51 [r22]
I180427 10:24:36.654906 62873 util/stop/stopper.go:471  quiescing; tasks left:
3      node.Node: batch
2      storage.intentResolver: processing intents
1      [async] sql.tableState: releasing descriptor lease
1      [async] kv.TxnCoordSender: heartbeat loop
I180427 10:24:36.654951 62873 util/stop/stopper.go:471  quiescing; tasks left:
3      node.Node: batch
2      storage.intentResolver: processing intents
1      [async] sql.tableState: releasing descriptor lease
W180427 10:24:36.654991 63412 storage/intent_resolver.go:388  [n1,s1] failed to cleanup transaction intents: could not GC completed transaction anchored at /Local/Range/Table/50/RangeDescriptor: result is ambiguous (server shutdown)
I180427 10:24:36.655010 62873 util/stop/stopper.go:471  quiescing; tasks left:
2      node.Node: batch
1      storage.intentResolver: processing intents
1      [async] sql.tableState: releasing descriptor lease
W180427 10:24:36.655184 63397 storage/intent_resolver.go:388  [n1,s1] failed to cleanup transaction intents: could not GC completed transaction anchored at /Table/SystemConfigSpan/Start: result is ambiguous (server shutdown)
I180427 10:24:36.655305 62873 util/stop/stopper.go:471  quiescing; tasks left:
1      [async] sql.tableState: releasing descriptor lease
W180427 10:24:36.655347 63311 internal/client/txn.go:531  failure aborting transaction: node unavailable; try another peer; abort caused by: result is ambiguous (server shutdown)
W180427 10:24:36.655382 63311 sql/lease.go:237  error releasing lease "51(\"foo\") ver=1:1524824963.246813111,0, refcount=0": result is ambiguous (server shutdown)
W180427 10:24:36.655613 63279 sql/lease.go:939  error: node unavailable; try another peer, not releasing lease: "51(\"foo\") ver=1:1524825015.479095056,0, refcount=0"
	lease_test.go:818: descriptor not acquired
```

--- a/pkg/sql/lease_test.go
+++ b/pkg/sql/lease_test.go
@@ -807,20 +807,17 @@ CREATE TABLE t.foo (v INT);
 		t.Fatal(err)
 	}
 
+	// This select can be retried in which case the descriptor gets reacquired.
 	if _, err := tx.Exec(`
 		SELECT * FROM t.foo;
 		`); err != nil {
 		t.Fatal(err)
 	}
 
-	// Descriptor has been acquired.
-	if atomic.LoadInt32(&fooAcquiredCount) != 1 {
-		t.Fatal("descriptor not acquired")
-	}
-
-	// Descriptor has not been released.
-	if atomic.LoadInt32(&fooReleaseCount) > 0 {
-		t.Fatal("released descriptor")
+	// Descriptor has been acquired one more time than it has been released.
+	aCount, rCount := atomic.LoadInt32(&fooAcquiredCount), atomic.LoadInt32(&fooReleaseCount)
+	if aCount != rCount+1 {
+		t.Fatalf("invalid descriptor acquisition counts = %d, %d", aCount, rCount)
 	}
 
 	if _, err := tx.Exec(
@@ -829,13 +826,13 @@ CREATE TABLE t.foo (v INT);
 		t.Fatal(err)
 	}
 
-	if atomic.LoadInt32(&fooAcquiredCount) > 1 {
-		t.Fatal("descriptor reacquired")
+	if cnt := atomic.LoadInt32(&fooAcquiredCount); cnt != aCount {
+		t.Fatalf("descriptor reacquired, %d != %d", cnt, aCount)
 	}
 
 	testutils.SucceedsSoon(t, func() error {
-		if atomic.LoadInt32(&fooReleaseCount) == 0 {
-			return errors.Errorf("didnt release descriptor")
+		if cnt := atomic.LoadInt32(&fooReleaseCount); cnt != aCount {
+			return errors.Errorf("didnt release descriptor, %d != %d", cnt, aCount)
 		}
 		return nil
 	})

